Source/theme.lua

```diff
@@ -0,0 +1,322 @@
+	end
+end
+
+function self:get_themes()
+	return self.themes
+end
+
+function self:get_active_themes()
+	return self.active_themes
+end
+
+function self:get_asset(name)
+	for i = #self.active_themes, 1, -1 do
+		local theme = self.active_themes[i]
+		local asset = self.themes[theme].assets[name]
+		if asset then
+			return asset
+		end
+	end
+	if allowdebug then
+		error("Asset " .. name .. " not found: " .. debug.traceback())
+	end
+end
+
+function self:draw(asset, x, y, r, sx, sy)
+	if type(asset) == "string" then
+		asset = self:get_asset(asset)
+	end
+
+	if (self:call("draw", asset, x, y, r, sx, sy)) then return end
+
+	if asset then
+		love.graphics.draw(asset.image, x, y, r, sx, sy)
+	end
+end
+
+function self:draw_nineslice(asset, x, y, x2, y2)
+	if type(asset) == "string" then
+		asset = self:get_asset(asset)
+	end
+
+	if (self:call("draw_nineslice", asset, x, y, x2, y2)) then return end
+
+	if asset then
+		local low_x = math.min(x, x2)
+		local low_y = math.min(y, y2)
+		local high_x = math.max(x, x2)
+		local high_y = math.max(y, y2)
+
+		-- draw nineslice, assuming all pieces are equal in size, so you can divide by 3
+		local w, h = asset.width, asset.height
+		local w2, h2 = w/3, h/3
+		if (math.floor(w2) ~= w2) or (math.floor(h2) ~= h2) then
+			error("Nineslice image " .. asset.name .. " is not divisible by 3.")
+		end
+
+		local draw_sprite_part = function(x, y, sx, sy, sw, sh)
+			local quad = self:get_quad(sx, sy, sw, sh, w, h)
+			love.graphics.draw(asset.image, quad, x, y)
+		end
+
+		-- draw the corners
+		draw_sprite_part(low_x, low_y, 0, 0, w2, h2)
+		draw_sprite_part(high_x - w2, low_y, w2*2, 0, w2, h2)
+		draw_sprite_part(low_x, high_y - h2, 0, h2*2, w2, h2)
+		draw_sprite_part(high_x - w2, high_y - h2, w2*2, h2*2, w2, h2)
+		-- draw the edges, looping through the middle
+		for i = low_x + w2, high_x - w2 * 2, w2 do
+			draw_sprite_part(i, low_y, w2, 0, w2, h2)
+			draw_sprite_part(i, high_y - h2, w2, h2*2, w2, h2)
+		end
+
+		for i = low_y + h2, high_y - h2 * 2, h2 do
+			draw_sprite_part(low_x, i, 0, h2, w2, h2)
+			draw_sprite_part(high_x - w2, i, w2*2, h2, w2, h2)
+		end
+
+		-- draw the middle, once again looping
+		for i = low_x + w2, high_x - w2 * 2, w2 do
+			for j = low_y + h2, high_y - h2 * 2, h2 do
+				draw_sprite_part(i, j, w2, h2, w2, h2)
+			end
+		end
+	end
+end
+
+function self:get_quad(x, y, w, h, image_width, image_height)
+	local key = x .. "_" .. y .. "_" .. w .. "_" .. h .. "_" .. image_width .. "_" .. image_height
+
+	if self.quads[key] then
+		return self.quads[key]
+	end
+
+	self.quads[key] = love.graphics.newQuad(x, y, w, h, image_width, image_height)
+	return self.quads[key]
+end
+
+function self:call(event, ...)
+	for i = #self.active_themes, 1, -1 do
+		local theme = self.active_themes[i]
+		if self.themes[theme].events and self.themes[theme].events[event] then
+			local result = self.themes[theme].events[event](...)
+			if result then
+				return result
+			end
+		end
+	end
+	return nil
+end
```

Comment:
Not sure what to do about this, because the event functions don't always have to be a complete replacement -- if theme 1 draws the nineslice tinted yellow, and theme 2 makes that nineslice a white png, then should theme 1's function still apply? Or should the white png in theme 2 override the function?

Reply:
So if theme 1 wasn't in the stack, then theme 2's white png would override theme 0's png, right? In which case I think if theme 1 _is_ in the stack, the white png in theme 2 should probably override the function, since theme 2 probably doesn't "expect" theme 1 to be overriding it. If themes 1 and 2 were the other way around (so the order is 0-2-1), then I guess theme 1 is "knowingly" applying the function to the image of whatever's below it (theme 2), assuming theme 1 doesn't already provide its own image.

What kind of usecases should we support? If an event function is "draw everything through yellow-colored glasses" then the user would want to put that theme at the top of the stack anyway, otherwise they're putting things in front of the glasses. If an event function is "draw a rectangle our own way without images" then putting another theme on top which provides its own png or function should override that. If an event function adds an extra subtitle at the bottom of the screen "you are mousing over a script box" when it doesn't interfere with how the box is drawn then you _would_ probably want that function to still run? But I wonder if we need functionality like that...

Maybe there could even be an extra argument to event functions, which tells the function it has already been overridden by something else higher in the stack (i.e. the thing may already have been completely drawn), but it can still do its thing if desired? But maybe we should just keep it simple. Or are there any other cases that are going to become a thing people won't understand? Because I'm just speaking from my imagination haha